Add fluent-plugin-remote-syslog gem

sqtran opened the pull request:
This is the work for adding the fluent-plugin-remote-syslog gem to openshift aggregated logging.  

There's one workaround here (7939ef3) that creates symlink to get around an issue https://github.com/openshift/origin-aggregated-logging/issues/604 where the gems cannot be found on the GEM_PATH after installation.

jcantrill commented on fluentd/generate_syslog_config.rb:
```diff
@@ -0,0 +1,54 @@
+  prefix = 'REMOTE_SYSLOG_'
+  group  = prefix + "HOST"
+
+  ENV.each_key.to_a.select { |s| s.start_with? group }.uniq.each do |k|
```
`ENV.keys.select { |s| s.start_with? group }.uniq.each do |k| `

jcantrill commented on fluentd/generate_syslog_config.rb:
```diff
@@ -0,0 +1,54 @@
+    # [0] = environment variable name
+    # [1] = syslog plugin config name
+    # [2] = default value, if any
+    vars = [
```
Seems this should only need to be done once and definitely more readable if it was a hash or collection of hashes
```
vars: {
  'HOST' : nil,
 'PORT' : 514
....
}
```

sqtran replied:
It's a little complicated.  It's an array of Hashes because it can handle the case if the user passes in multiple ENV variables to configure mulitple syslog servers.  Any ENV variables that start with the prefix "REMOTE_SYSLOG" will be grouped based on their suffix.

For example, you could set REMOTE_SYSLOG_HOST=192.168.1.1 and REMOTE_SYSLOG_HOST_2=192.168.1.2 as ENV variables.  This script will build two fluentd \<store\> configurations.  Group suffixes are any characters after the string "REMOTE_SYSLOG_HOST" - as that is the single required field to build any configuration at all.

So above, the two groups would be "" and "_2".  That's why it's an array of Hashes.

jcantrill commented on fluentd/generate_syslog_config.rb:
```diff
@@ -0,0 +1,54 @@
+    ]
+    t = k.dup
+    t.slice! group
+    vars.each { |r| r[2] = ENV[prefix + r[0] + t] unless !ENV[prefix + r[0] + t] }
```
there is an opportunity using the above suggestion to rebuild the env using 'lower' and 'map' possibly

jcantrill commented on fluentd/generate_syslog_config.rb:
```diff
@@ -0,0 +1,54 @@
+<store>
+@type syslog_buffered
+"
+     r.each { |v|  c << v[1]  << ' ' << v[2] << "\n" unless !v[2] }
```
```
r.each { |v|  c <<  "#{v[1]} #{v[2]}\n" unless !v[2] }
```

sqtran replied:
Ah yes, I forgot about string interpolation.  This makes it a lot more readable.

jcantrill commented on fluentd/generate_syslog_config.rb:
```diff
@@ -0,0 +1,54 @@
+"
+  end
+
+  File.open(file_name, 'w') { |f| f.write(c) }
```
I'm not certain but do we need to explicitly close the file handle here?

sqtran replied:
No, when passing an (inline) block to File, it gets closed automatically at the end.

richm commented on fluentd/run.sh:
```diff
@@ -217,6 +217,10 @@ else
     umount /var/lib/docker/containers/*/shm || :
 fi
 
+if [[ $USE_REMOTE_SYSLOG ]] ; then
```
```
if [ "${USE_REMOTE_SYSLOG:-false}" = true ] ; then
```

richm commented on fluentd/Dockerfile.centos7:
```diff
@@ -29,6 +29,7 @@ RUN rpmkeys --import file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7 && \
 RUN mkdir -p ${HOME} && mkdir -p /etc/fluent/plugin && \
     gem install -N --conservative --minimal-deps --no-ri \
      'activesupport:<5' \
+     'public_suffix:<3' \
```
You'll have to rebase to pick up this fix.

richm commented on test/remote-syslog.sh:
```diff
@@ -0,0 +1,76 @@
+
+os::log::info Test 1, expecting generate_syslog_config.rb to have created configuration file
+
+fpod=`get_running_pod fluentd`
```
`$( get_running_pod fluentd )`

richm commented on test/remote-syslog.sh:
```diff
@@ -0,0 +1,76 @@
+os::log::debug "$( oc label node --all logging-infra-fluentd=true --overwrite=true )"
+os::cmd::try_until_text "oc get pods -l component=fluentd" "^logging-fluentd-.* Running "
+
+fpod=`get_running_pod fluentd`
```
`$( get_running_pod fluentd )`

richm commented:
Just a few minor comments.  Please address them, then rebase to latest master, and squash these into a single commit.

sqtran commented:
Does this look right?  I merged my branch with upstream/master, rebased and fast-forwarded.  Then I made my code review changes, rebased it into a single commit, then pushed.  My only changes are in 5946c88.

richm commented:
This doesn't look right.  Not sure why there are so many commits in this PR.  What exactly did you do?

sqtran commented:
I had to rewrite history to clean up the mess you saw.  This branch is now rebased with upstream/master, and includes the requested code review fixes.

richm commented:
This is what I usually do to rebase and squash:
```
pwd
~/origin-aggregated-logging
git branch my-working-branch-with-many-commits
git checkout master
git pull
git checkout my-working-branch-with-many-commits
git rebase master
# fix up any conflicts, etc.
git rebase -i master
# use pick on the first commit
# use squash on all subsequent commits
# merge the commit messages into one commit message
# now my-working-branch-with-many-commits has only 1 commit
git push --force myrepo my-working-branch-with-many-commits
# had to force push to my branch because I rewrote history
```

sqtran commented:
@richm Oh!  So you want ALL my work as a single commit?

richm commented:
@sqtran yes

sqtran commented:
All squashed now

richm commented:
[test]

richm commented:
[test]

richm commented:
[test]

richm commented:
[merge]

richm commented:
curator [merge] flake

jcantrill commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(5) Include and support fluentd-syslog-plugin[logging]](https://trello.com/c/LCte3O0n/469-5-include-and-support-fluentd-syslog-pluginlogging)
